feat: Add URL params support

lfmnovaes opened the pull request:
- Replaced localStorage manipulation with support for saving the passive skill tree state directly to URL parameters.

Note: This PR also includes my previous changes related to the 'Clear Search' button. #86

Preview:
![image](https://github.com/user-attachments/assets/8fd294e7-f7cd-4aec-811a-26ec772c694f)

lfmnovaes commented:
Updated to also store the Ascendancy and their nodes, if the user picks another Ascendancy, only (but all of them) the Ascendancy nodes are unselected.

lfmnovaes commented:
> This is awesome! This remind me of my embedded dev days :) From reading it seems that this approach could break if there was a new node added to the tree. I was originally thinking of using a simple JSON -> base64 or something like that. Of course it would make the string way longer

I understand your concern, but since the IDs are unique and the code works by encoding/decoding from an array of IDs, adding new nodes to the tree shouldn’t cause any issues. I’ve also tested it before and after the changes to the .json file, including scenarios where completely new nodes were introduced (like those from Ascendancy) and it was completely compatible.

My first approach was indeed to be simpler with something like base64, but the URL was getting pretty big. This second approach results in much smaller URLs, and the trade-off is a small package and a few calculations.

Feel free to test it out, and let me know if you come across any bugs!

marcoaaguiar commented:
The nodes are ordered keystones -> notables -> smalls
Let's say we have 20 keystones, then 500 notables and 1000 smalls. If select the 1 first notable that would be entry 21 of the array of bits.
If a new keystone is added, now the 21 entry is the last keystone. 
Am I wrong?

I guess this could be solved by keeping keystones, notables, smalls and ascendency nodes separately

lfmnovaes commented:
> The nodes are ordered keystones -> notables -> smalls Let's say we have 20 keystones, then 500 notables and 1000 smalls. If select the 1 first notable that would be entry 21 of the array of bits. If a new keystone is added, now the 21 entry is the last keystone. Am I wrong?
> 
> I guess this could be solved by keeping keystones, notables, smalls and ascendency nodes separately

That's correct, if a completely new keystone (not previously mapped) arrives, it will point to the wrong nodes. And yes, might be an issue in the long run, didn't consider that. I'm going to try a different approach then.

meatwallace commented:
probably worth asking just how long a URL is if you take the simplified approach of just stringifying the raw selected node list and forgo any optimisation at all.

i pulled your branch and simplified it to a bare bones implementation like:

```ts
const compressedNodeIDS = LZString.compressToEncodedURIComponent(JSON.stringify(selectedNodes));
```

a URL with 132 selected nodes looks like:


```
http://localhost:5173/?ascendancy=bloodmage&nodes=NoIgggcgjCA04QExwQZhZALCpUbyUwFYdFV8QlEBOUxZAxKW%2BAZQHZEAOFD1d3u0zY27Lj1HUGIVlygA2XlwAMEmV3JL502URHrUitl0xrZ8o%2BpVKNWiqyn2u7J8UHsWM9kXSiovmXkPXnk9EKCQnxDmEOp9PgCOKGVYxKCzIPj5Z0iBNnl%2BSMtWAoyuYqC0uViddmVPViIpaJJGZWKidgDoLpwodksenU7hml4ievHlVpkiBXHUGdZMVEThBuEzFZ1iNWhUaWgxtkxk3lP7U51UajyZVHE%2B%2BZPUFJe9qEwdz-PbNh97Ho7o0DucxOdNCdlPYbhRoD82Pw3vd6LxVktXjoaHDvDguFiiHtwWxEBNeIhsnizGQlogVuTJmxmMjWJ8zP0dP0lgpuXUGSzEMpgYLisxPBBCbwFIkFDDtGj5A0HjMjpdlPFMIyZJrulBBWiTAqzIY4XgUgBdIA
```

which obviously isn't as "nice" as having a shorter URL, but this isn't that crazy long for this type of thing - seen plenty of implementations that produce URLs that look like that. it's also not something that's human decipherable in the end anyway, so as far as usability goes, nothing changes.

my suggestion would be to take the path of least resistance here.

lfmnovaes commented:
Thanks for the feedbacks.
Changed the compression approach to be more reliable in the future, and fixed conflicts

Preview:
![image](https://github.com/user-attachments/assets/51fea033-4d4d-4190-a240-f89bd04c858a)

marcoaaguiar commented:
@meatwallace do you mind reviewing the PR?

meatwallace commented on src/routes/+page.svelte:
```diff
@@ -46,43 +47,54 @@
-		selectedNodes = selectedNodes.filter((nodeId) => nodes[nodeId].class === ascendancy);
+	$: if (selectedAscendancy !== prevSelectedAscendancy) {
+		// Remove Ascendancy nodes from selectedNodes when changing between ascendancies
+		selectedNodes = selectedNodes.filter((id) => !id.startsWith('A'));
```
i'd revert this line - it's not a huge deal but comparing the actual data is more robust than checking the string prefix. probably wouldn't cause issues but it feels brittle.

lfmnovaes replied:
Not experienced with Svelte either, done this way to store the previous Ascendancy selection and then trigger a small cleanup when they don't match.

meatwallace commented on src/routes/+page.svelte:
```diff
@@ -46,43 +47,54 @@
 
-	function handleAscendancyChange(ascendancy: string) {
-		selectedNodes = selectedNodes.filter((nodeId) => nodes[nodeId].class === ascendancy);
+	$: if (selectedAscendancy !== prevSelectedAscendancy) {
```
i'm not particularly experienced with Svelte - do we need to break the ascendency change handler fn out into imperative code? seems like Svelte was handling comparing current/prev values for us prior to this.

lfmnovaes replied:
This behavior is intentional. The previous implementation would wipe out all selected nodes when changing the Ascendancy (we already have a button for that). With this approach, only the Ascendancy related nodes are unselected.

meatwallace replied:
rather than blocking getting the feature in, it's probably safe to just send it as is and someone can do a followup PR to extract logic/organise things in a uniform manner, if it's even deemed worth it (given the scope of the project)

meatwallace reviewed (commented):
haven't approved or explicitly requested changes as both bits of feedback aren't truly blocking, it's more or less just an opinion. with all the logic sitting in the page component as it is i think it's not a bad idea to encapsulate the logic where possible, though. 

haven't pulled & ran the branch but the code otherwise lgtm. good feature :)

meatwallace commented:
feedback to my feedback makes sense, lgtm.

marcoaaguiar commented:
Agreed merging
